src/streamlink/plugins/ustreamtv.py

```diff
@@ -79,11 +80,32 @@ def cluster(self, cluster):
         self._cluster = cluster
         self.reconnect()
 
+    @staticmethod
+    def parse_proxy_url(purl):
+        proxy_options = {}
+        if purl:
+            p = urlparse(purl)
+            proxy_options['proxy_type'] = p.scheme
+            proxy_options['http_proxy_host'] = p.hostname
+            if p.port:
+                proxy_options['http_proxy_port'] = p.port
+            if p.username:
+                proxy_options['http_proxy_auth'] = (unquote_plus(p.username), unquote_plus(p.password or ""))
+        return proxy_options
+
     def connect(self):
-        log.debug("Connecting to {0}".format(self.host))
+        proxy_options = self.parse_proxy_url(self._proxy_url)
+        if proxy_options.get('http_proxy_host'):
+            log.debug("Connecting to {0} via proxy ({1}://{2}:{3})".format(self.host,
+                                                                           proxy_options.get('proxy_type') or "http",
+                                                                           proxy_options.get('http_proxy_host'),
+                                                                           proxy_options.get('http_proxy_port') or 80))
+        else:
+            log.debug("Connecting to {0}".format(self.host))
         self._ws = websocket.create_connection(self.host,
```

Comment:
`websocket._exceptions.WebSocketProxyException:` should maybe be handled for invalid proxies

```
[plugin.ustreamtv][debug] Connecting to ws://xxx via proxy (http://xxx)
Traceback (most recent call last):
  File "streamlink/vv/bin/streamlink", line 11, in <module>
    load_entry_point('streamlink', 'console_scripts', 'streamlink')()
  File "streamlink/src/streamlink_cli/main.py", line 1037, in main
    handle_url()
  File "streamlink/src/streamlink_cli/main.py", line 577, in handle_url
    streams = fetch_streams(plugin)
  File "streamlink/src/streamlink_cli/main.py", line 457, in fetch_streams
    sorting_excludes=args.stream_sorting_excludes)
  File "streamlink/src/streamlink/plugin/plugin.py", line 325, in streams
    ostreams = list(ostreams)
  File "streamlink/src/streamlink/plugins/ustreamtv.py", line 465, in _get_streams
    api.connect()
  File "streamlink/src/streamlink/plugins/ustreamtv.py", line 108, in connect
    **proxy_options)
  File "streamlink/vv/lib/python3.7/site-packages/websocket/_core.py", line 514, in create_connection
    websock.connect(url, **options)
  File "streamlink/vv/lib/python3.7/site-packages/websocket/_core.py", line 223, in connect
    options.pop('socket', None))
  File "streamlink/vv/lib/python3.7/site-packages/websocket/_http.py", line 122, in connect
    sock = _tunnel(sock, hostname, port, auth)
  File "streamlink/vv/lib/python3.7/site-packages/websocket/_http.py", line 293, in _tunnel
    "failed CONNECT via proxy status: %r" % status)
websocket._exceptions.WebSocketProxyException: failed CONNECT via proxy status: 403

  File "streamlink/vv/lib/python3.7/site-packages/websocket/_http.py", line 293, in _tunnel
    "failed CONNECT via proxy status: %r" % status)
websocket._exceptions.WebSocketProxyException: failed CONNECT via proxy status: 503

  File "streamlink/vv/lib/python3.7/site-packages/websocket/_http.py", line 293, in _tunnel
    "failed CONNECT via proxy status: %r" % status)
websocket._exceptions.WebSocketProxyException: failed CONNECT via proxy status: 400
```

Reply:
Yep - I’ll take a look :)

Reply:
this `exception` is **not important** when used with `--http-proxy` it will fail earlier